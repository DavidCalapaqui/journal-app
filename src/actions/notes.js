import { db } from "../firebase/firebase-config";
import { collection, addDoc, updateDoc, doc, deleteDoc } from "firebase/firestore";
import {types} from '../types/types';

import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";



export const startNewNote = () =>{
    
    
    return async (dispatch, getState) =>{
        const {uid} = getState().auth;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const doc = await addDoc(collection(db, `${uid}/journal/notes`), newNote);
            //console.log('DOC REF: ', doc);

            dispatch(activeNote(doc.id, newNote));
            dispatch(addNewNote(doc.id, newNote));
        } catch (error){
            console.error('ERROR AL INTENTAR GRABAR UNA COLECCION: ',error);
        } 
       
    }
}

export const startLoadingNotes = (uid) =>{
    return async(dispatch)=>{
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}

export const activeNote = (id, note) =>({
    type: types.notesActive,
    payload:{
        id, 
        ...note
    }
})

export const addNewNote = (id, note)=>({
    type: types.notesAddNew,
    payload:{
        id,
        ...note
    }
})

export const setNotes = (notes) =>({
    type: types.notesLoad,
    payload: notes
})


export const startSaveNote = (note) =>{
    return async (dispatch, getState)=>{
        const {uid} = getState().auth;

        if(!note.url){
            delete note.url;
        }

        const noteToFirestore = {...note};
        delete noteToFirestore.id;
        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
        try {
             await updateDoc(noteRef, noteToFirestore);
        } catch (error) {
            console.error('ERROR AL ACTUALIZAR NOTA: ',error);
        }

        dispatch(refreshNote(note.id, note));
        Swal.fire('Saved', note.title, 'success')
    }
}

export const refreshNote = (id, note) =>({
    type: types.notesUpdated,
    payload:{
        id, 
        note:{
            id,
            ...note
        }
    }
})


export const startUploading = (file) =>{

    return async (dispatch,getState ) =>{
        const {active:activeNote} = getState().notes;
        Swal.fire({
            title: 'Uploading',
            text: 'Please wait...',
            allowOutsideClick: false,
            onBeforeOpen: ()=>{ 
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote))
        Swal.close();
    }
}


export const startDeleting = (id) =>{
    return async (dispatch, getState)=>{
        const uid = getState().auth.uid;

        const noteRef = doc(db, `${uid}/journal/notes/${id}`);

        try {
            await deleteDoc(noteRef);
            dispatch(deleteNote(id));
            Swal.fire('Note removed', 'The note has been removed on database', 'success')

        } catch (error) {
            Swal.fire('Error', error.message, 'error')
        }

    }
}

export const deleteNote = (id) =>({
    type: types.notesDelete,
    payload: id
})

//limpiar las notas del usuario en sesión cuando cierre sesión
export const notesLogout =()=>({
    
    type: types.notesLogoutCleaning
})