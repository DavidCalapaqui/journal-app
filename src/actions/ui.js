import { types } from "../types/types";

//ACCIONES PARA ERRORES EN CAMPOS
export const setError = (err) =>({
    type: types.uiSetError,
    payload: err
})

export const removeError = () =>({
    type: types.uiRemoveError,
})

//ACCIONES PARA CARGANDO
export const startLoading = () =>({
    type: types.uiStartLoading,
})

export const finishLoading = () =>({
    type: types.uiFinishLoading,
})