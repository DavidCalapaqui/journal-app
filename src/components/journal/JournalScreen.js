import React from 'react'
import { useSelector } from 'react-redux'
import { NoteScreen } from '../notes/NoteScreen'
import { NothingSelected } from './NothingSelected'
import { Sidebar } from './Sidebar'

//import { NothingSelected } from './NothingSelected'
export const JournalScreen = () => {
    

    //jalo el active del state.notes
    const {active} = useSelector(state => state.notes);

    
    //si tengo una nota seleccionada la muestro si no la pantalla de sleccione una
    return (
        <div className='journal__main-content animate__animated animate__fadeIn animate__faster'>
            <Sidebar />
        
            <main>
                {
                    (active)
                    ? <NoteScreen />
                    : <NothingSelected />
                }
                             
            </main>
        </div>
    )
}
