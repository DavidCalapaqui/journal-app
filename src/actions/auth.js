
import {getAuth, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { finishLoading, startLoading } from './ui';
import { notesLogout } from './notes';

//accion
export const startLoginEmailPassword = (email, password) =>{
    //retorna un callback 
    return (dispatch)=>{
        //dispatch startLogin
        dispatch(startLoading());

        const auth = getAuth();
        return signInWithEmailAndPassword(auth, email, password)
        .then(async ({user})=>{
            dispatch(login(user.uid, user.displayName));
            
            dispatch(finishLoading());
            //dispatch stopLoggin
            //console.log('EMAIL & PASS - USER SIGIN: \n',user.displayName, user.uid);
        }).catch(e=>{
            Swal.fire('Error', e.message, 'error' );
            }
        );
    }

}

export const startRegisterWithEmailPasswordName = (email, password, name) =>{
    return(dispatch)=>{
        const auth = getAuth();
        createUserWithEmailAndPassword(auth, email, password)
        .then(async({user})=>{ 
            //dispatch(login(user.uid, user.displayName));
            await updateProfile(user, {displayName: name});
            dispatch(user.uid, user.displayName);
            console.log(user);
        })
        .catch(e=>{
            Swal.fire('Error', e.message, 'error' );
        })
    }

}

//accion
export const startGoogleLogin = () =>{
    return (dispatch)=>{
       const auth = getAuth();
       signInWithPopup(auth, googleAuthProvider)
       .then(({user})=>{
            console.log('GOOGLE - USER SIGIN: \n', user.displayName, user.uid);
           dispatch(login(user.uid, user.displayName));
       })
    }
}
//accion
export const login = (uid, displayName) =>{
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

//acción asíncrona a firebase
export const startLogout = () =>{
    const auth = getAuth();

    return async (dispatch)=>{
        await auth.signOut();
        dispatch(logout()); 
        //disparar accion para limpiar notas
        dispatch(notesLogout());
    }
}

export const logout = () =>({
    type: types.logout
});