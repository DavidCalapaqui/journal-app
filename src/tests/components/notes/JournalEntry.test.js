import React from 'react';
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';

import { activeNote } from '../../../actions/notes';

import { JournalEntry } from '../../../components/journal/JournalEntry';




const middlewares = [thunk];
const mockStore = configureStore(middlewares);

//jest.mock('../../../actions/notes', ()=>({
//    activeNote: jest.fn(),
//}))

const initState = {}

//simulacion del store
let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
    id: 10,
    date: 0,
    title: 'Hola',
    body: 'mundo',
    url: 'https://somewhere.com/photos/delfo.jpg'
}

const wrapper = mount(
    
    <Provider store={store}>
            <JournalEntry {...note}/>
    </Provider>
        
);

describe('Pruebas en JournalEntry', ()=>{

    test('Debe de mostrarse correctamente', () => { 
        expect(wrapper).toMatchSnapshot();
        
    })

    test('Debe de actiar la nota', () => { 
        wrapper.find('.journal__entry').prop('onClick')();
        expect(store.dispatch).toHaveBeenCalled();
        expect(store.dispatch).toHaveBeenCalledWith(
            activeNote(note.id, {...note})
        );

     })


})