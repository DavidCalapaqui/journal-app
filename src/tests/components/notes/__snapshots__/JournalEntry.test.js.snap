// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en JournalEntry Debe de mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://somewhere.com/photos/delfo.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hola
    </p>
    <p
      className="journal__entry-content"
    >
      mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
